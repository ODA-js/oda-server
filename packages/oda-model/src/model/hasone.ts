import clean from '../lib/json/clean';
import { EntityReference } from './entityreference';
import { HasOneInput, HasOneStorage } from './interfaces';
import { RelationBase } from './relationbase';

export class HasOne extends RelationBase {
  protected $obj: HasOneStorage;

  get hasOne(): EntityReference {
    return this.$obj.hasOne;
  }

  get ref(): EntityReference {
    return this.$obj.hasOne;
  }

  constructor(obj: HasOneInput) {
    super(obj);
  }

  public updateWith(obj: HasOneInput) {
    if (obj) {
      super.updateWith(obj);
      const result = { ...this.$obj };

      this.setMetadata('storage.single', true);
      this.setMetadata('storage.stored', false);
      this.setMetadata('storage.embedded', obj.embedded || false);
      this.setMetadata('verb', 'HasOne');

      let $hasOne = obj.hasOne;

      let hasOne;
      if ($hasOne) {
        hasOne = new EntityReference($hasOne);
        if (!hasOne.backField) {
          hasOne.backField = 'id';
        }
      }

      result.hasOne_ = new EntityReference($hasOne).toString();
      result.hasOne = hasOne;

      this.$obj = result;
      this.initNames();
    }
  }

  // it get fixed object
  public toObject(): any {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      hasOne: props.hasOne ? props.hasOne.toString() : undefined,
    });
  }

  // it get clean object with no default values
  public toJSON(): any {
    let props = this.$obj;
    let res = super.toJSON();
    return clean({
      ...res,
      hasOne: props.hasOne_,
    });
  }
}
