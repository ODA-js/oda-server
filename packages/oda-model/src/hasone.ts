import {
  IRelationBase,
  RelationBasePersistence,
  RelationBaseMetaInfo,
  RelationBaseInput,
  RelationBaseInternal,
  RelationBase,
  RelationBaseOutput,
} from './relationbase';
import { IEntityRef, EntityReference } from './entityreference';
import { merge } from 'lodash';
import { assignValue, Nullable } from './model';

export interface HasOnePersistence extends RelationBasePersistence {
  single: boolean;
  stored: boolean;
  embedded: boolean;
}

export interface IHasOneRelation
  extends IRelationBase<
    HasOneMetaInfo,
    HasOneInput,
    HasOnePersistence,
    HasOneOutput
  > {
  hasOne: IEntityRef;
}

export interface HasOneMetaInfo
  extends RelationBaseMetaInfo<HasOnePersistence> {}

export interface HasOneInternal
  extends RelationBaseInternal<HasOneMetaInfo, HasOnePersistence> {
  hasOne: IEntityRef;
}

export interface HasOneInput
  extends RelationBaseInput<HasOneMetaInfo, HasOnePersistence> {
  hasOne: string;
}

export interface HasOneOutput
  extends RelationBaseOutput<HasOneMetaInfo, HasOnePersistence> {
  hasOne: string;
}

const defaultMetaInfo = {
  verb: 'HasOne',
  persistence: {
    single: true,
    stored: false,
    embedded: false,
  },
};
const defaultInternal = {};
const defaultInput = {};

export class HasOne extends RelationBase<
  HasOneMetaInfo,
  HasOneInput,
  HasOneInternal,
  HasOnePersistence,
  HasOneOutput
> {
  get hasOne(): IEntityRef {
    return this.$obj.hasOne;
  }

  get ref(): IEntityRef {
    return this.$obj.hasOne;
  }

  constructor(inp: HasOneInput) {
    super(merge({}, defaultInput, inp));
    this.metadata_ = merge({}, defaultMetaInfo, this.metadata_);
    this.$obj = merge({}, defaultInternal, this.$obj);
  }

  public updateWith(input: Nullable<HasOneInput>) {
    super.updateWith(input);

    assignValue<HasOneMetaInfo, HasOneInput, boolean>({
      src: this.metadata_,
      input,
      inputField: 'embedded',
      effect: (src, value) => (src.persistence.embedded = value),
    });

    assignValue<HasOneInternal, HasOneInput, string>({
      src: this.$obj,
      input,
      field: 'hasOne',
      effect: (src, value) => {
        src.hasOne = new EntityReference(value);
        if (!src.hasOne.backField) {
          src.hasOne.backField = 'id';
        }
      },
    });
  }

  // it get fixed object
  public toObject(): HasOneOutput {
    return merge({}, super.toObject(), {
      hasOne: this.hasOne.toString(),
    } as Partial<HasOneOutput>);
  }
}
