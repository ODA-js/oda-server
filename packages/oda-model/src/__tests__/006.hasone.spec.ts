import 'jest';
import { HasOne, hasOneDefaultMetaInfo, hasOneDefaultInput } from '../hasone';
import { merge } from 'lodash';

describe('hasOne', () => {
  it('default', () => {
    const res = new HasOne({ hasOne: 'A#' });
    expect(res.metadata).toMatchObject(hasOneDefaultMetaInfo);
    expect(res.toObject()).toMatchObject(
      merge({ hasOne: 'id@A#id' }, hasOneDefaultInput),
    );
    expect(res.modelType).toBe('HasOne');
    expect(res.verb).toBe('HasOne');
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
