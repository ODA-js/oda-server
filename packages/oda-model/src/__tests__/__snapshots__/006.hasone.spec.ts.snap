// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hasOne default: toObject 1`] = `
Object {
  "entity": undefined,
  "field": undefined,
  "hasOne": "id@A#id",
  "metadata": Object {
    "name": Object {
      "full": "HasOne",
      "normal": "",
      "short": "",
    },
    "persistence": Object {
      "embedded": false,
      "single": true,
      "stored": false,
    },
    "verb": "HasOne",
  },
  "name": "HasOne",
  "opposite": undefined,
}
`;
